Image memory optimization to prevent OOM crashing and other improvements

iAndrea opened the pull request:
Following Issue: https://github.com/fatihak/InkyPi/issues/426

deftdawg commented:
This is the way! :smile: 

Some other things I was considering tinkering with:
- Saving image to a temp file and streaming it to resize, rather than downloading it in mem and then resizing it
- Centralizing download and re-scaling of images into the parent "plugin" class, since this is an issue that is likely to occur for unsplash, wpotd, apod, image_url, image_folder, image_upload, ai_image(maybe)...

iAndrea commented:
@deftdawg Got it!

- temp file -> I'm already working on it 🎉 
- centralization -> I'll need more time but sure, I already though about it since it is definitely a common problem throughout the project since low ram is low ram everywhere 🙃

iAndrea commented:
@deftdawg I'm pretty close to completion, just a bunch of tests and I'll push the changes for a central image management :thumbnail

iAndrea commented:
So here we are with the next update: "Centralize image processing with adaptive memory optimization"

**Problem**
Multiple plugins were independently downloading and processing images with no memory optimizations, causing potential crashes on Raspberry Pi Zero 2W (512MB RAM) when handling high-resolution images (4000x6000+). Each plugin had duplicate image loading code with inconsistent error handling.

**Solution**
Implemented centralized _AdaptiveImageLoader_ utility that automatically optimizes based on device capabilities:

1. Low-resource devices (Pi Zero, <1GB RAM):
 - Streams downloads to temp files (8KB chunks, not full image in RAM)
 - Uses PIL draft mode for 80-90% memory reduction during JPEG decode
 - Two-stage resize (thumbnail → LANCZOS) for huge images
 - Explicit garbage collection after each stage
2. High-performance devices (Pi 3/4, ≥1GB RAM):
 - Fast in-memory processing
 - Full-resolution decode with LANCZOS (highest quality)
 - No disk I/O overhead

The new utility `src/utils/image_loader.py` is integrated into BasePlugin so then:
 - All plugins now have self.image_loader available
 - Simple API: .from_url(), .from_file(), .from_bytesio()

Migrated 8 plugins:
✅ apod
✅ image_url
✅ wpotd
✅ image_folder
✅ image_upload
✅ ai_image
✅ comic
✅ unsplash
Each has been tested tested on Pi Zero 2W except for apod, image_folder, ai_image.

Logging:
 - Added comprehensive INFO/DEBUG/ERROR logging to all migrated plugins
 - Consistent format matching Unsplash implementation
 - Shows device detection, processing strategy, and optimization steps

iAndrea commented:
@fatihak Can you take a look into this PR? I think it would improve quite a lot the project stability for low-memory devices 🙏

iAndrea commented:
Due to the introduction of the new `image_album` plugin with support for Immich, this new plugin is subject to the same problems we already talked about. In fact I tried to retrieve a picture of mine with resolution 6000x4000 and it crashed. I've already updated that plugin to support the new features:
- Replaced Direct HTTP Requests with Centralized Session
- Integrated AdaptiveImageLoader
- Enhanced Logging
  - Added comprehensive logging throughout the plugin (matching patterns from unsplash, apod, image_url)
  - Detailed logging for configuration, dimensions, and processing steps
  - Better error messages with context
- Improved Error Handling
  - Fixed bug in get_album_id() where it would fail if album not found
  - Added check for empty asset lists
  - Better exception messages with more context
- Match Orientation awareness
  - Selects only images matching the device settings according to exif (dimensions + orientation flags)
  - For images with slightly different ratio compared to the device, using crop-to-fill instead of pad-to-fit for optimal display

deftdawg commented:
Not to minimize the other excellent work done here, but I would rename this PR ticket to `Image memory optimization to prevent OOM crashing and other improvements`, I think keeping InkyPi from dying from OOM when it loads big images is the most important aspect from a community perspective - most of the problems people currently report with Inky dying are very likely caused by OOM that this resolves...   @fatihak

fatihak commented on src/plugins/wpotd/wpotd.py:
```diff
@@ -44,26 +43,42 @@ def generate_settings_template(self) -> Dict[str, Any]:
+        dimensions = (max_width, max_height)
+
+        # Use adaptive loader if shrink-to-fit is enabled
+        if settings.get("shrinkToFitWpotd") == "true":
```
This can just be simplified to self._download_image(picurl, dimensions=dimensions, resize= settings.get("shrinkToFitWpotd") == "true")

fatihak commented on src/plugins/image_album/settings.html:
```diff
@@ -21,6 +21,25 @@
+        </div>
+        <!-- Hidden input that actually submits the value -->
+        <input type="hidden" id="orientationFilter" name="orientationFilter" value="true">
+        <small style="display: block; margin-top: 5px; color: #666; font-size: 0.9em;">
```
This feature isn't needed and isn't related to the memory optimization

fatihak commented on src/plugins/image_album/image_album.py:
```diff
@@ -1,74 +1,239 @@
+
+        return detailed_assets
+
+    def filter_assets_by_orientation(self, assets: list[dict], dimensions: tuple[int, int]) -> list[dict]:
```
Not needed

fatihak commented on src/plugins/unsplash/settings.html:
```diff
@@ -45,17 +45,32 @@
         </select>
     </div>
+
+    <div class="form-group">
```
Not sure why this is needed if the code forces downsizing on low-resource devices. I think we should go with regular size on low-resource devices and full otherwise, without this option in the UI.

fatihak commented:
Thanks for the changes. Also just realized when testing that the behavior of the "Scale to Fit" option is changed. We don't want to stretch images with this option, it should just scale the image, keeping the aspect ratio, to fit the dimensions of the screen. 
Here is a sample image I used:
<img width="155" height="454" alt="test_image" src="https://github.com/user-attachments/assets/93201032-764a-4973-88f5-26eac50793ff" />

Image displayed previously with 'Scale to Fit" option enabled:
![IMG_1608](https://github.com/user-attachments/assets/09d8e969-0e1f-4f00-b159-c401c2795219)


Image displayed now with 'Scale to Fit' option enabled:
![IMG_1609](https://github.com/user-attachments/assets/d3d46ce0-bdd3-484f-9c39-886146a0039d)

iAndrea commented on src/plugins/image_album/settings.html:
```diff
@@ -21,6 +21,9 @@
         </div>
         <!-- Hidden input ensures 'true' is always sent when unchecked -->
         <input type="hidden" name="padImage" value="true">
+        <small style="display: block; margin-top: 5px; color: #666; font-size: 0.9em;">
```
Sure 👍

fatihak commented:
I saw this issue in the image upload plugin, but seems like you only updated the image album plugin
